appservice/src/SiteWrapper.ts

```diff
@@ -116,7 +127,60 @@ export class SiteWrapper {
         this.log(outputChannel, 'Deployment completed.');
     }
 
-    private async waitForDeploymentToComplete(kuduClient: KuduClient, outputChannel: vscode.OutputChannel, pollingInterval: number = 5000): Promise<void> {
+    public async localGitDeploy(fsPath: string, client: WebSiteManagementClient, outputChannel: vscode.OutputChannel): Promise<DeployResult | undefined> {
+        const kuduClient: KuduClient = await this.getKuduClient(client);
+        const yes: string = 'Yes';
+        const pushReject: string = localize('localGitPush', 'Push rejected due to Git history diverging. Force push?');
+        const scmType: string = 'LocalGit';
+
+        const [publishCredentials, config]: [User, SiteConfigResource] = await Promise.all([
+            this.getWebAppPublishCredential(client),
+            this.getSiteConfig(client)
+        ]);
+
+        if (config.scmType !== scmType) {
+            // SCM must be set to LocalGit prior to deployment
+            const scmUpdate: string | undefined = await this.updateScmType(client, config, scmType);
+            if (scmUpdate !== scmType) {
+                // if the new config scmType doesn't equal LocalGit, user either canceled or there was an error
+                return undefined;
+            }
+        }
+        // credentials for accessing Azure Remote Repo
+        const username: string = publishCredentials.publishingUserName;
+        const password: string = publishCredentials.publishingPassword;
+        const remote: string = `https://${username}:${password}@${this._gitUrl}`;
+        const localGit: git.SimpleGit = git(fsPath);
+        try {
+
+            const status: git.StatusResult = await localGit.status();
+            if (status.files.length > 0) {
+                const uncommit: string = localize('localGitUncommit', '{0} uncommitted change(s) in local repo "{1}"', status.files.length, fsPath);
+                vscode.window.showWarningMessage(uncommit);
+            }
+            await localGit.push(remote, 'HEAD:master');
+        } catch (err) {
+            // tslint:disable-next-line:no-unsafe-any
+            if (err.message.indexOf('spawn git ENOENT') >= 0) {
+                await this.showInstallPrompt();
+                return undefined;
+            } else if (err.message.indexOf('error: failed to push') >= 0) { // tslint:disable-line:no-unsafe-any
+                const input: string | undefined = await vscode.window.showErrorMessage(pushReject, yes);
+                if (input === 'Yes') {
+                    await (<(remote: string, branch: string, options: object) => Promise<void>>localGit.push)(remote, 'HEAD:master', { '-f': true });
+                    // Ugly casting neccessary due to bug in simple-git. Issue filed:
+                    // https://github.com/steveukx/git-js/issues/218
+                } else {
+                    return undefined;
+                }
+            } else {
+                return undefined;
```

Comment:
Not all errors should return undefined. You should call `throw err` here. I was just saying don't throw a _custom_ error type because then you also have to export that error type